### ISSUE TITLE
%%timeit could use local scope.

### ISSUE THREAD

**Issue by Carreau** [ipython/ipython#10556]
You can't use timeit in a function (which is useful to test variable  inputs of several function doing the same things, and plot graphs.

I believe a well placed `@needs_local_scope` and modification to timeit should allow that.

**Comment by adityausathe** [comment]
In current scenario, the `inputSplitter` and `inputTransformer` treat `%%timeit ... ` as line-magic and replaces it with a call to`%sget_ipython().magic(%r)`  function,  which just ignores the line. Do we need to distinguish local scoped `%%timeit... ` while parsing input or is there any other way?

**Comment by takluyver** [comment]
I'm not sure what you mean about the parsing - `%%timeit` definitely works.

You can't have a cell magic in a local scope because it has to be a whole cell - only the line magic form can be in a local scope.

**Comment by adityausathe** [comment]
I was working on it and I was able to get it work(ran some basic tests). Only problem was, I had to enter two blank lines to complete the cell.
My implementation considers all the remaining lines in the cell starting from `%%timeit`, as magic body. Please, tell me if I'm doing wrong here.

**Comment by takluyver** [comment]
> Only problem was, I had to enter two blank lines to complete the cell.
> My implementation considers all the remaining lines in the cell starting from %%timeit, as magic body. 

I think that's normal for a cell magic. Is it any different from `%%timeit` in master?

**Comment by Carreau** [comment]
I was likely refering to the line magic `%timeit` not the cell magic. I tried to make a %timeit, in a function in a for loop. For example:

```
def fib(n):
    return 1 if n < 2 else fib(n-1)+fib(n-2)

def timefib(P):
    t = []
    for i in range(P):
        res = %timeit -o fib(i)
        t.append(res)
    return t
```

And this timeit would need local scope.

I believe that what I was referring to and  inadvertantly have put 2 % sign.

**Comment by adityausathe** [comment]
ooh. :disappointed:

**Comment by adityausathe** [comment]
I worked on it and was able to narrow down the needed changes. We need to pass `local_ns` to `exec(code, self.shell.user_ns, local_ns)` call. But, it was still giving `NameError`. I think that is due to custom `inner` function
```
timeit_ast_template = ast.parse('def inner(_it, _timer):\n'
                                        '    setup\n'
                                        '    _t0 = _timer()\n'
                                        '    for _i in _it:\n'
                                        '        stmt\n'
                                        '    _t1 = _timer()\n'
                                        '    return _t1 - _t0\n')
```
But, `%timeit` is working fine in local scope, if I merge local namespace(`local_ns`) and global namespace and pass it as global to `exec`. I am not getting this? Please, guide me on this.

**Comment by takluyver** [comment]
I think passing a single combined local and global namespace to `exec()` *should* work, but I haven't carefully thought through the possible ramifications.

@Carreau for future reference, namespace handling is probably a bit fiddly and delicate to invite new contributors to work on. That's not a criticism of @adityausathe - you seem to be getting into it well - I just spotted that this bug was tagged 'Needs contributor'.

**Comment by Carreau** [comment]
I though it would be simpler. I didn't though about the nested function and exec. I agree that this one is a bit tough.

**Comment by adityausathe** [comment]
Should we go with merging `local_ns` into `global_ns`? I think, the problem with that would be the variables having both scopes (same name) . If `stmt` in `def inner()` wants global value of a overriden variable, then, it might create trouble there.

**Comment by adityausathe** [comment]
Since, I'm passing merged namespaces, this patch fails in following scenario -
```
In [1]: def fib(n):
   ...:     global j
   ...:     j = 'me'
   ...:     return 1 if n < 2 else fib(n-1)+fib(n-2)
   ...:
In [2]: def timefib(P):
   ...:     t = []
   ...:     j= 99
   ...:     for i in range(P):
   ...:         res = %timeit -o fib(i)
   ...:         t.append(res)
   ...:     return t
   ...: 
In [3]: j = 'hey'

In [4]: timefib(3)
105 ns ± 0.0458 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
106 ns ± 0.0306 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
326 ns ± 0.415 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
Out[4]: 
[<TimeitResult : 105 ns ± 0.0458 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)>,
 <TimeitResult : 106 ns ± 0.0306 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)>,
 <TimeitResult : 326 ns ± 0.415 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)>]

In [5]: j
Out[5]: 'hey'
```
The `j` should have been changed `'me'` but since, we are restoring back the duplicate variables present in `local_ns`  after execution of `%timeit`, the changed value `global j; j = 'me'` gets neglected.

The `j` in `local_ns` and then invoking `global j` inside `%timeit` body, causes this problem.
`{'local_ns' : { 't' : [], 'P' : 3, 'j' : 99}`

**Comment by Carreau** [comment]
> The j should have been changed 'me' but since, we are restoring back the duplicate variables present in local_ns after execution of %timeit, the changed value global j; j = 'me' gets neglected.

That's seem ok to me. Time-it being limitted in what it can do is fine – IMHO it will always be. You could have `fib` do a `j +=1`, and have an inconsistency between systems about how many times it increment `j`.

**Comment by takluyver** [comment]
Can you check what value `fib()` sees for `j` before it changes it, as well? I'm a bit concerned that it might see `j = 99` from `timefib()`, rather than the global `j = 'hey'`.

If that's OK, I think I agree with @Carreau that not modifying the global j is acceptable. I think it's already the case that j won't be modified by timeit code like this:

```
j = 1
%timeit j = 2
```

So

**Comment by adityausathe** [comment]
> Can you check what value `fib()` sees for `j` before it changes it, as well? I'm a bit concerned that it might see `j = 99` from `timefib()`, rather than the global `j = 'hey'`.

Yes, you are right. The genuine global var `j` is getting overwritten. 
But, I have got another way of solving this, How about nested calls to `exec()`. Inner call will be made from `inner()` and we could pass `local_ns` to `inner()` first and then `exec()` will use them to execute `stmt`. 
But, I haven't thought about its repercussions on the execution time measurement.

**Comment by Carreau** [comment]
> Yes, you are right. The genuine global var j is getting overwritten.
> But, I have got another way of solving this, How about nested calls to exec(). Inner call will be made from inner() and we could pass local_ns to inner() first and then exec() will use them to execute stmt.
> But, I haven't thought about its repercussions on the execution time measurement.

I think we can stop where we are for the complexity. It's already really good, and I'm worries about exec'ing in the function. You _can_ execute the inside of a function without creating a new local scope (see https://github.com/ipython/ipython/pull/10390), and let's not forget:

> Simple is better than complex.
> Complex is better than complicated.

I'm going to merge the relevant PR and we can revisit if/when we have complaints.

**Comment by takluyver** [comment]
> Yes, you are right. The genuine global var j is getting overwritten.

That's unfortunate. I would have been inclined to hold off from merging the PR in that case. Hopefully not too many people will hit it, though, because it only affects using `%timeit` inside a function, which I think is fairly unusual.

> But, I have got another way of solving this, How about nested calls to exec(). Inner call will be made from inner() and we could pass local_ns to inner() first and then exec() will use them to execute stmt.

That's a decent idea, but I think we probably shouldn't pursue it. Timeit goes to some lengths to eliminate as much code as possible between the time measurements and the code to execute, so that the timings are as accurate as possible.
